### ISSUE TITLE
ItemID.Sets.Deprecated Bool Forces Mod Recipes to Not Appear

### ISSUE THREAD

**Issue by localMisanthrope** [bluemagic123/tModLoader#2487]
### Version

Latest 1.4-stable

### OS

Windows

### Platform

Steam

### This bug affects

Mod capability as a Modder

### Description

`ItemID.Sets.Deprecated["insert random item here"] = true;`

This line in particular _works_ as intended, however it simultaneously forces modded recipes to never appear in the crafting UI.

Oddly enough, this does not include vanilla recipes, only modded. 

### Log File

Log File:

- Log file is irrelevant as it provides no information.

### Steps to reproduce

1. Create a `GlobalItem`

2. Insert the following lines;

```
public static readonly int[] depArray = { ItemID.CopperBar, ItemID.CopperShortsword... };

public override void SetDefaults(Item item) {
for (int i = 0; i < depArray.length; i++){
ItemID.Sets.Deprecate[depArray[i]] = true;
}
}
```

3. Build and reload, enter world, and watch as your modded recipes magically disappear!

### Expected Behavior

Expected Behavior:

- Items should be "deprecated", or removed from all spaces.

- Items no longer visible and their own recipes are removed.

### Actual Behavior

Actual Behavior:

- Expected behavior executes as intended.

- Forces modded recipes to never appear in the crafting UI.

### Reproduction frequency

Reproduction Frequency:

- Occurs without fail every time this code is executed.

### Additional Information

Speculation:

- Possible that deprecation forces empty items and recipe spaces to somehow "invalidate" modded recipes, maybe by referencing values that don't exist? What's odd is nothing in logs shows anything regarding this, and modded recipes (as far as debug yields) DO get set properly on load. 

Going to see if I can't find a workaround until `ItemCreationContext` finally gets it's own "spawn in" contexts for my own testing. That or I'll just tell this crafting system to kick it to the curb 😑

**Comment by Ishigh1** [comment]
Well first in your example you're clearly not using the set as intended, you're setting a static flag that should be set in a SetStaticDefaults or any methods that is run once, at launch in a SetDefaults, whenever an item is created
Also it has nothing to do with modded recipes or whatever, if you set for example the torch as deprecated in ModSystem.PostSetupContent, the first recipe in Terraria which is usually creating a torch doesn't exist anymore as its result is air.
The recipe UI is lazy and stop if a recipe it sees is air, so in our case, it will stop at the first recipe an not show ANY recipe.

So in conclusion, the Deprecate array is bad, there are few ways to solve this issue : 
- Not allow recipes with empty results in the array
- Disable those recipes and ask the UI to ignore and not stop when seeing a disabled recipe with an air result
- NOT change anything about Deprecated, maybe even privatize it and ask modders to disable the recipes themselves and put a SetDefaults(0) in GlobalItem.SetDefaults() for the chosen items

**Comment by Chicken-Bones** [comment]
Option 3 is the best. Making items automatically turn into air seems like a bad idea. Mods should just remove crafting recipes for them. Sure, you can `SetDefaults(0)` in a `GlobalItem.SetDefaults` but that's likely to break a lot of stuff so I wouldn't do that.

**Comment by Ishigh1** [comment]
So switch to internal or not ?

**Comment by JavidPack** [comment]
I use `ItemID.Sets.Deprecated` in some of my mod logic while iterating over all items in the game, making it internal would be a pain point. Maybe we need to just add some comments saying that it won't do what you think, but we are open to discussing it as a feature in a future tmod version.

**Comment by localMisanthrope** [comment]
> I use `ItemID.Sets.Deprecated` in some of my mod logic while iterating over all items in the game, making it internal would be a pain point. Maybe we need to just add some comments saying that it won't do what you think, but we are open to discussing it as a feature in a future tmod version.

Honestly I think we just need a clear cut method or something for this. I want to remove items without it destroying my brain.

**Comment by JavidPack** [comment]
If there is a strong enough demand for this, it is definitely something that could be implemented. (I've dreamed of a mod that deprecates all existing tiles and items and has a full new experience) Someone would need to think it through and figure out all the places it would need to have effect.

For example, what happens if a recipe only has a deprecated ingredient, will the recipe persist with no ingredients and be free?
Will loot drops from the bestiary be properly removed if a mod attempts to deprecate an item? When in the loading process would mods need to deprecate items? Network desync warnings?

I'm sure there are more that I haven't thought of yet.

**Comment by localMisanthrope** [comment]
> If there is a strong enough demand for this, it is definitely something that could be implemented. (I've dreamed of a mod that deprecates all existing tiles and items and has a full new experience) Someone would need to think it through and figure out all the places it would need to have effect.
> 
> For example, what happens if a recipe only has a deprecated ingredient, will the recipe persist with no ingredients and be free? Will loot drops from the bestiary be properly removed if a mod attempts to deprecate an item? When in the loading process would mods need to deprecate items? Network desync warnings?

I uh... am already making a total conversion mod. Well, working on one, that is.

This has been a pain for me since I can't properly make space for my content and absolutely dump out the old. Problem is I don't consider myself knowledgable enough to create a method of that scale in order to properly compensate for everything. At least not yet.

Pretty sure you _might_ be able to compensate for air recipes in some way. Getting the recipe UI to work in tandem however might be the only issue. Bestiary data shoulnd't be an inherent issue since editing vanilla bestiary entries isn't hard (last I checked, anyway). Should be done at the beginning of mod load to prevent content from getting in the way I would think.

**Comment by Chicken-Bones** [comment]
@JavidPack has only started opening the can of worms. The solution is to make items unobtainable, not remove them from the game.

Maybe that's what `Deprecated` should do, or maybe we should make a new set, but it should:
- Change the tooltip color
- _maybe_ make the item as plain as possible in `SetDefaults` (non-functional)
- Be something other mods can check for before displaying in cheat-sheet

 If we're considering deprecating vanilla items or items from other mods then maybe we want to do some stuff in tML for convenience (but not that we _have_ to cover every case).
- Disable recipes which make or use it
- Remove/disable NPC drop rules which produce it
